central/auth/datastore/datastore_impl_test.go

```diff
@@ -0,0 +1,240 @@
+//go:build sql_integration
+
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	pgStore "github.com/stackrox/rox/central/auth/datastore/internal/store/postgres"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/errox"
+	"github.com/stackrox/rox/pkg/postgres/pgtest"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stretchr/testify/suite"
+)
+
+func TestAuthDatastorePostgres(t *testing.T) {
+	suite.Run(t, new(datastorePostgresTestSuite))
+}
+
+type datastorePostgresTestSuite struct {
+	suite.Suite
+
+	ctx       context.Context
+	pool      *pgtest.TestPostgres
+	store     pgStore.Store
+	datastore DataStore
+}
+
+func (s *datastorePostgresTestSuite) SetupTest() {
+	s.ctx = sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.Access),
+		),
+	)
+
+	s.pool = pgtest.ForT(s.T())
+	s.Require().NotNil(s.pool)
+
+	s.store = pgStore.New(s.pool.DB)
+	s.datastore = New(s.store)
+}
+
+func (s *datastorePostgresTestSuite) TearDownTest() {
+	s.pool.Teardown(s.T())
+	s.pool.Close()
+}
+
+func (s *datastorePostgresTestSuite) TestAddConfig() {
+	testCases := []struct {
+		config *storage.AuthMachineToMachineConfig
+		err    error
+	}{
+		{
+			config: &storage.AuthMachineToMachineConfig{},
+			err:    errox.InvalidArgs,
+		},
+		{
+			config: nil,
+			err:    errox.InvalidArgs,
+		},
+		{
+			config: &storage.AuthMachineToMachineConfig{
+				Id: "some-id",
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &storage.AuthMachineToMachineConfig{
+				TokenExpirationDuration: "1h",
+			},
+		},
+		{
+			config: &storage.AuthMachineToMachineConfig{
+				TokenExpirationDuration: "1s",
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &storage.AuthMachineToMachineConfig{
+				TokenExpirationDuration: "24h1s",
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &storage.AuthMachineToMachineConfig{
+				TokenExpirationDuration: "1h",
+				Type:                    storage.AuthMachineToMachineConfig_GENERIC,
+			},
+			err: errox.InvalidArgs,
+		},
+		{
+			config: &storage.AuthMachineToMachineConfig{
+				TokenExpirationDuration: "1h",
+				Type:                    storage.AuthMachineToMachineConfig_GENERIC,
+				IssuerConfig: &storage.AuthMachineToMachineConfig_Generic{
+					Generic: &storage.AuthMachineToMachineConfig_GenericIssuer{Issuer: "something"}},
+			},
+		},
+	}
```

Comment:
Also, a case where `id` is set in the input object would demonstrate that `Add*` does not care about `id` value on the input and always sets a new one.

Although, like I hinted in other comments, it feels these tests more belong to testing the overall service. My idea is that here you validate a behavior that ultimately is a behavior of the service. Sure, the logic can be validated at this level too and is perhaps the right thing to do in multiple situations (including this one). However, ultimately the serivce is what the caller interacts with, and this service is not too complex. Moving tests to the service layer means: we care about the service behavior and prove/validate that through tests; the fact that the service uses certain constellation of stores+datastores is merely an implementation detail.  
The combined behavior of service and its dependencies will be validated by tests.

Again, moving tests to the service is a subjective thing and is not always right. For instance, in the situation with data validation logic: you likely don't want to exhaustively check every case of invalid data by calling service methods because that will require a bunch of redundant code. Such thing as data validation should be tested independently and service tests can just check that the service does not forget to perform data validation where you expect it to.

Similarly, with the conversion logic. It should be tested as independent unit (like you already have it) and then you can depend on that knowing it's reliable because you tested it.

At this point you may say that I contradict myself. What I'm saying is that I see the service (its business logic) + conversion and datastore + store + postgres all _currently_ exist to serve one purpose and this is exposed through the service. In my opinion, the end result becomes more flexible in situations _when you want to replace or heavily refactor storage layer_ when tests are placed at the service level. I.e. I suggest to downgrade "conversion and datastore" and underlying things to be an implementation detail.

Reply:
Moved most tests to the service layer.

What I kept within the datastore layer was testing the correct behavior of the foreign key since this seemed like an "implementation" detail on the service layer, not necessarily on the datastore layer however.